### PR TITLE
ocamlPackages.janeStreet{,_0_9_0}: join the ocamlPackages fix point, allowing overriding to work as expected

### LINKED ISSUE #75485: Ocaml packages are not individually overridable
I am facing the problem that I want to use Ocaml 4.08, but the `base` package in nixpkgs (1.11.1) is incompatible with 4.08.

I wanted to use a local overlay to bump it to 1.13, but it would not get picked up by other packages, and the reason seems to be that the package set in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/ocaml-modules/janestreet/default.nix is a simple `rec`, and does _not_ take part in the global fixed-point of nix packages. Shouln’t this file take `base` as a parameter and _not_ be a `rec` set?

Is that intentionally or was it just an oversight? @vbgl?

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Fix long standing `ocamlPackages` issue #75485: The `janeStreet` package set is maintained as a sub set of `ocamlPackages`  and merged into it using `liftJaneStreet` in a second step. This is not an issue in itself, however inner-janestreet dependencies are done by using a `rec` set in the respective nix expressions. This means that `janeStreet` packages are oblivious to overrides being done in `ocamlPackages` to other `janeStreet` packages and use the original derivation regardless.

This is now fixed, see the commit message of ff40bbe for an example of this.

Another issue arising due to this is that `janeStreet` packages are both available via `ocamlPackages.package` and `ocamlPackages.janeStreet.package`. This creates an issue with overriding as overriding one the two attributes will always be ignored, as we only can use either `ocamlPackages` or `ocamlPackages.janeStreet` as the `self` fix point to get dependencies from.  Due to `liftJaneStreet` it feels more natural to use the top level attribute. To prevent confusion in the future we add a warning when using `ocamlPackages.janeStreet` instead of the top level attributes (2f413da).

cc @nomeata (original reporter) @vbgl

GitHub thingy: Resolves #75485.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Seems like I have to revert af2dbfd because `ofborg` doesn't allow any remaining `trace` messages when evaluating (which is fair).

The warnings stem from these bits:

```nix
    ppx_optcomp =
      if lib.versionOlder "4.03" ocaml.version
      then janeStreet.ppx_optcomp
      else callPackage ../development/ocaml-modules/janestreet/ppx-optcomp.nix {};
```

Alternatively we could either

* Move that into `liftJaneStreet`
* Remove the legacy versions

Which also begs the question: Should we start removing some of the _old_ janestreet stuff? Most packages in `janeStreet_0_9_0` for example fail to build by now because its dependencies use Dune 2 and their jbuild invocations have become incompatible.

**Comment by sternenseemann** [comment]
I've had a look at those conditional version changes and they are not that big of a problem:

* Most are irreperably failing to evaluate and can be removed
* We can move the rest of the conditionals that do evaluate and build into the `janeStreet` 0.9 set

File: pkgs/development/ocaml-modules/janestreet/0.12.nix [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
+{ self
 , openssl
 }:
 
-rec {
+with self;
+
+{
 
   ocaml-compiler-libs = janePackage {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/janestreet/0.12.nix:9]:
```suggestion
with self; {
  ocaml-compiler-libs = janePackage {
```

File: pkgs/development/ocaml-modules/janestreet/0.14.nix [hunk 2]
```diff
@@ -1,30 +1,11 @@
...
-, ounit
 , zstd
 }:
 
-rec {
+with self;
+
+{
 
   accessor = janePackage {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/janestreet/0.14.nix:10]:
```suggestion
with self; {
  accessor = janePackage {
```

File: pkgs/development/ocaml-modules/janestreet/default.nix [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
+{ self
+, openssl
 }:
 
-rec {
+with self;
+
+{
 
   ocaml-compiler-libs = janePackage {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/janestreet/default.nix:9]:
```suggestion
with self; {
  ocaml-compiler-libs = janePackage {
```

File: pkgs/development/ocaml-modules/janestreet/old.nix [hunk 4]
```diff
@@ -1,8 +1,32 @@
...
+    num
+    ;
+
+in
+
+with self;
+
+{
 
   # Jane Street packages, up to ppx_core
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/janestreet/old.nix:31]:
```suggestion
in
with self; {
  # Jane Street packages, up to ppx_core
```

**Review by SuperSandro2000 — COMMENTED** [review]
small nit

**Comment by sternenseemann** [comment]
> small nit

makes the diff larger and more awkward to read imo

**Comment by sternenseemann** [comment]
@ofborg eval

**Pull request merged**